test/widgets/action_sheet_test.dart

```diff
@@ -0,0 +1,229 @@
+              accountId: eg.selfAccount.id,
+              child: MessageListPage(narrow: narrow)))));
+
+      // global store, per-account store, and message list get loaded
+      await tester.pumpAndSettle();
+
+      // request the message action sheet
+      final messageContent = tester.widget<MessageContent>(
+        find.byWidgetPredicate((Widget widget) => widget is MessageContent && widget.message.id == message.id));
+      await tester.longPress(find.byWidget(messageContent));
+      // sheet appears onscreen; default duration of bottom-sheet enter animation
+      await tester.pump(const Duration(milliseconds: 250));
+    }
+
+    ComposeBoxController? findComposeBoxController(WidgetTester tester) {
+      return tester.widgetList<ComposeBox>(find.byType(ComposeBox))
+        .single.controllerKey?.currentState;
+    }
+
+    Widget? findQuoteAndReplyButton(WidgetTester tester) {
+      return tester.widgetList(find.byIcon(Icons.format_quote_outlined)).singleOrNull;
+    }
+
+    void prepareRawContentResponseSuccess(PerAccountStore store, {
+      required Message message,
+      required String rawContent,
+    }) {
+      // Prepare fetch-raw-Markdown response
+      // TODO: Message should really only differ from `message`
+      //   in its content / content_type, not in `id` or anything else.
+      (store.connection as FakeApiConnection).prepare(json:
+        GetMessageResult(message: eg.streamMessage(contentMarkdown: rawContent)).toJson());
+    }
+
+    void prepareRawContentResponseError(PerAccountStore store) {
+      final fakeResponseJson = {
+        'code': 'BAD_REQUEST',
+        'msg': 'Invalid message(s)',
+        'result': 'error',
+      };
+      (store.connection as FakeApiConnection).prepare(httpStatus: 400, json: fakeResponseJson);
+    }
+
+    /// Simulates tapping the quote-and-reply button in the message action sheet.
+    ///
+    /// Checks that there is a quote-and-reply button.
+    Future<void> tapQuoteAndReplyButton(WidgetTester tester) async {
+      final quoteAndReplyButton = findQuoteAndReplyButton(tester);
+      check(quoteAndReplyButton).isNotNull();
+      await tester.tap(find.byWidget(quoteAndReplyButton!));
+    }
+
+    void checkLoadingState(PerAccountStore store, ComposeContentController contentController, {
+      required TextEditingValue valueBefore,
+      required Message message,
+    }) {
+      check(contentController).value.equals((ComposeContentController()
+        ..value = valueBefore
+        ..insertPadded(quoteAndReplyPlaceholder(store, message: message))
+      ).value);
+      check(contentController).validationErrors.contains(ContentValidationError.quoteAndReplyInProgress);
+    }
+
+    void checkSuccessState(PerAccountStore store, ComposeContentController contentController, {
+      required TextEditingValue valueBefore,
+      required Message message,
+      required String rawContent,
+    }) {
+      final c = ComposeContentController()
```

Comment:
Yeah; hmm. I guess it's a _builder_ for an expected value, which it provides at `.value`, rather than anything we want to make expectations on directly. My preference would actually be to not store it in a variable at all, and in an earlier draft I had this:

```dart
      check(contentController).value.equals(
        (ComposeContentController()
          ..value = valueBefore
          ..insertPadded(quoteAndReply(store, message: message, rawContent: rawContent))
        ).value);
```

But that code doesn't actually produce the `TextEditingValue` that we want; in particular, it hasn't set its `selection` to be the last index of the text. To do that, we need to read what that last index is, and I didn't find a way to do that with the `..` "cascade notation".